File: cpp_src/zig_random.cpp
#include "zig_random.h"

unsigned long splitmix64rng(unsigned long &seed) {
    unsigned long ssplit = (seed += 0x9E3779B97F4A7C15UL);
    ssplit = (ssplit ^ (ssplit >> 30)) * 0xBF58476D1CE4E5B9UL;
    ssplit = (ssplit ^ (ssplit >> 27)) * 0x94D049BB133111EBUL;
    return ssplit ^ (ssplit >> 31);
}

unsigned long left_rotate(unsigned long src, int magnitude) {
    return (src << magnitude) | (src >> (64 - magnitude));
}

Xoshiro512StarStarRandom::Xoshiro512StarStarRandom() {
    this->initialize_state(30984686UL);
}

Xoshiro512StarStarRandom::Xoshiro512StarStarRandom(unsigned long seed) {
    this->initialize_state(seed);
}

void Xoshiro512StarStarRandom::initialize_state(unsigned long seed){
    /*hash the seed to set the intial state*/
    this->_s0 = splitmix64rng(seed);
    this->_s1 = splitmix64rng(seed);
    this->_s2 = splitmix64rng(seed);
    this->_s3 = splitmix64rng(seed);
    this->_s4 = splitmix64rng(seed);
    this->_s5 = splitmix64rng(seed);
    this->_s6 = splitmix64rng(seed);
    this->_s7 = splitmix64rng(seed);
    return; 
};

unsigned long Xoshiro512StarStarRandom::next_ulong(void) {
    /*new set of 64 random bits*/
    unsigned long result = left_rotate(this->_s1 * 5UL, 7) * 9UL;

    /*Update the state*/
    unsigned long tmp = this->_s1 << 11;
    this->_s2 ^= this->_s0;
    this->_s5 ^= this->_s1;
    this->_s1 ^= this->_s2;
    this->_s7 ^= this->_s3;
    this->_s3 ^= this->_s4;
    this->_s4 ^= this->_s5;
    this->_s0 ^= this->_s6;
    this->_s6 ^= this->_s7;
    this->_s6 ^= tmp;
    this->_s7 = left_rotate(this->_s7, 21);

    return result;
};

unsigned int Xoshiro512StarStarRandom::next_uint(void) {
    /*New set of 32 random bits*/
    return (unsigned int)this->next_ulong();
}

double Xoshiro512StarStarRandom::rand_0I1E(void) {
    /*Return a number between [0, 1)*/
    return this->next_ulong() * this->incr_double;
}

double Xoshiro512StarStarRandom::rand_0E1I(void) {
    /*Return a number between (0, 1]*/
    return this->rand_0I1E() + this->incr_double;
}

File: c#_src/zig_random.cs
public class Utils {
    public static ulong splitmix64rng(ref ulong seed) {
        ulong ssplit = (seed += 0x9E3779B97F4A7C15UL);
        ssplit = (ssplit ^ (ssplit >> 30)) * 0xBF58476D1CE4E5B9UL;
        ssplit = (ssplit ^ (ssplit >> 27)) * 0x94D049BB133111EBUL;
        return ssplit ^ (ssplit >> 31);
    }

    public static ulong left_rotate(ulong src, int magnitude) {
        return (src << magnitude) | (src >> (64 - magnitude));
    }
}

public class Xoshiro512StarStarRandom {
    private ulong _s0, _s1, _s2, _s3, _s4, _s5, _s6, _s7;
    private const double incr_double = 1.0 / (1UL << 53);

    public Xoshiro512StarStarRandom(){
        initialize_state(30984686UL);
    }

    public Xoshiro512StarStarRandom(ulong _seed){
        initialize_state(_seed);
    }

    public void initialize_state(ulong _seed){
        _s0 = Utils.splitmix64rng(ref _seed);
        _s1 = Utils.splitmix64rng(ref _seed);
        _s2 = Utils.splitmix64rng(ref _seed);
        _s3 = Utils.splitmix64rng(ref _seed);
        _s4 = Utils.splitmix64rng(ref _seed);
        _s5 = Utils.splitmix64rng(ref _seed);
        _s6 = Utils.splitmix64rng(ref _seed);
        _s7 = Utils.splitmix64rng(ref _seed);
    }

    public ulong next_ulong(){
        /*Get 64 bits*/
        ulong result = Utils.left_rotate(_s1 * 5UL, 7) * 9UL;

        /*Update the state*/
        ulong tmp = _s1 << 11;
        _s2 ^= _s0;
        _s5 ^= _s1;
        _s1 ^= _s2;
        _s7 ^= _s3;
        _s3 ^= _s4;
        _s4 ^= _s5;
        _s0 ^= _s6;
        _s6 ^= _s7;
        _s6 ^= tmp;
        _s7 = Utils.left_rotate(_s7, 21);

        return result;
    }

    public uint next_uint() {
        /*next 32 random bits*/
        return (uint)next_ulong();
    }

    public double rand_0I1E() {
        /*Random double between [0, 1)*/
        return (double)next_ulong() * Xoshiro512StarStarRandom.incr_double;
    }

    public double rand_0E1I() {
        /*Random double between (0, 1]*/
        return rand_0I1E() + Xoshiro512StarStarRandom.incr_double;
    }
}